### PR TITLE
test(e2e): move universal gateway tests to e2e_env

### PR DESCRIPTION
**Pull request by michaelbeaumont**
### Checklist prior to review

- [x] Link to docs PR or issue --
- [x] Link to UI issue or PR --
- [x] Is the [issue worked on linked][1]? -- #4520 
- [x] The PR does not hardcode values that might break projects that depend on kuma (e.g. "kumahq" as a image registry) --
- [x] The PR will work for both Linux and Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [x] Unit Tests --
- [x] E2E Tests --
- [x] Manual Universal Tests --
- [x] Manual Kubernetes Tests --
- [x] Do you need to update [`UPGRADE.md`](/UPGRADE.md)? --
- [x] Does it need to be backported according to the [backporting policy](/CONTRIBUTING.md#backporting)? --
- [x] Do you need to explicitly set a [`> Changelog:` entry here](/CONTRIBUTING.md#submitting-a-patch)?

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

File: test/e2e_env/universal/gateway/gateway.go [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+			Install(echoServerApp(mesh, "echo-server", "echo-service", "universal")).
+			Install(echoServerApp(mtlsMesh, "echo-server-mtls", "echo-service", "universal")).
+			Install(GatewayProxyUniversal(mesh, "gateway-proxy")).
+			Install(GatewayProxyUniversal(mtlsMesh, "gateway-proxy-mtls")).
+			Install(ExternalServerUniversal(mesh))
+
+		Expect(setup.Setup(env.Cluster)).To(Succeed())
+
+		Expect(
+			env.Cluster.GetKumactlOptions().KumactlApplyFromString(MkGateway("edge-gateway", mesh, false, "example.kuma.io", "echo-service", gatewayPort)),
```
Comment (jakubdyszkiewicz) [inline test/e2e_env/universal/gateway/gateway.go:53]: nit: can we chain it in ClusterSetup with `YamlUniversal(MkGateway...`?

File: test/e2e_env/universal/gateway/gateway.go [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+			return cluster.DeployApp(
+				WithArgs([]string{"test-server", "echo", "--port", "8080", "--instance", "external-echo"}),
+				WithName("external-echo-"+mesh),
+				WithMesh(mesh),
+				WithoutDataplane(),
+				WithVerbose())
+		}
+	}
+
+	// DeployCluster creates a universal Kuma cluster using the
```
Comment (jakubdyszkiewicz) [inline test/e2e_env/universal/gateway/gateway.go:36]: we do not create Kuma cluster here, we use a cluster that was already created

File: test/e2e_env/universal/gateway/gateway.go [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+		It("should proxy simple HTTP requests", func() {
+			proxySimpleRequests(env.Cluster, "universal",
+				GatewayAddressPort("gateway-proxy-mtls", mtlsGatewayPort), "example-mtls.kuma.io")
+		})
+
+		// In mTLS mode, only the presence of TrafficPermission rules allow services to receive
+		// traffic, so removing the permission should cause requests to fail. We use this to
+		// prove that mTLS is enabled
+		It("should fail without TrafficPermission", func() {
+			proxyRequestsWithMissingPermission(env.Cluster, mtlsMesh,
```
Comment (jakubdyszkiewicz) [inline test/e2e_env/universal/gateway/gateway.go:94]: I think this helper is overkill. It's used once and the test logic is basically squeezed there
```
// when traffic permission is removed
// then traffic is blocked
```

can we inline it?

File: test/e2e_env/universal/gateway/gateway.go [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+				GatewayAddressPort("gateway-proxy-mtls", mtlsGatewayPort), "example-mtls.kuma.io")
+		})
+	})
+
+	Context("when targeting an external service", func() {
+		BeforeAll(func() {
+			// The suite-level JustBeforeEach adds a default route to the mesh echo server.
```
Comment (jakubdyszkiewicz) [inline test/e2e_env/universal/gateway/gateway.go:101]: we don't have JustBeforeEach anymore

File: test/e2e_env/universal/gateway/gateway.go [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+		// In mTLS mode, only the presence of TrafficPermission rules allow services to receive
+		// traffic, so removing the permission should cause requests to fail. We use this to
+		// prove that mTLS is enabled
+		It("should fail without TrafficPermission", func() {
+			proxyRequestsWithMissingPermission(env.Cluster, mtlsMesh,
+				GatewayAddressPort("gateway-proxy-mtls", mtlsGatewayPort), "example-mtls.kuma.io")
+		})
+	})
+
+	Context("when targeting an external service", func() {
```
Comment (jakubdyszkiewicz) [inline test/e2e_env/universal/gateway/gateway.go:99]: nit: can we move `ExternalServerUniversal` and install `ExternalServerUniversal` in this context?

File: test/e2e_env/universal/gateway/gateway.go [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+conf:
+  http:
+    requests: 5
+    interval: 10s
+`, mesh)),
+			).To(Succeed())
+		})
+
+		It("should be rate limited", func() {
+			proxyRequestsWithRateLimit(env.Cluster,
```
Comment (jakubdyszkiewicz) [inline test/e2e_env/universal/gateway/gateway.go:190]: could we inline this? It's only used once

File: test/e2e_env/universal/gateway/gateway.go [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+
+			// Create the TLS secret containing the self-signed certificate and corresponding private key.
+			Expect(
+				env.Cluster.GetKumactlOptions().KumactlApplyFromString(fmt.Sprintf(`
+type: Secret
+mesh: %s
+name: example-kuma-io-certificate
+data: %s
+	`, mesh, payload)),
+			).To(Succeed())
```
Comment (jakubdyszkiewicz) [inline test/e2e_env/universal/gateway/gateway.go:244]: nits:
1) I think we should create secrets first
2) Can we just override existing gateway? less code
3) `env.Cluster.Install(YamlUniversal(` is less verbose than `env.Cluster.GetKumactlOptions().KumactlApplyFromString(` and more visible around e2e codebase

File: test/e2e_env/universal/gateway/gateway.go [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+
+			Expect(
+				env.Cluster.GetKumactlOptions().KumactlList("meshgateways", mesh),
+			).To(ContainElement("edge-https-gateway"))
```
Comment (jakubdyszkiewicz) [inline test/e2e_env/universal/gateway/gateway.go:248]: I don't think we need this check.

File: test/e2e_env/universal/gateway/gateway.go [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+				addr,
+				client.Resolve(addr, GatewayAddress("gateway-proxy")),
+			)
+
+			proxySecureRequests(
+				env.Cluster,
+				"universal",
+				GatewayAddressPort("gateway-proxy", 8081),
+			)
+		})
```
Comment (jakubdyszkiewicz) [inline test/e2e_env/universal/gateway/gateway.go:265]: I think we have 2 test cases here
```
1) should proxy https requests
1) should proxy https requests via hostname
```
also I'd use DescribeTabe and inline `proxySecureRequests`, but that's my opinion

**Review by jakubdyszkiewicz — COMMENTED** [review]
nice!

File: test/e2e_env/universal/gateway/gateway.go (hunk 6 above)
Comment (michaelbeaumont) [reply test/e2e_env/universal/gateway/gateway.go:190]: inlined and rewritten with `CollectFailure`.

File: test/e2e_env/universal/gateway/gateway.go (hunk 7 above)
Comment (michaelbeaumont) [reply test/e2e_env/universal/gateway/gateway.go:244]: 1) ok
2) Sure though a bit conflicted on making everything here order-dependent. My gut feeling says HTTP and HTTPS tests shouldn't depend on each other.
3) Definitely, I didn't actually realize there was an `Install` method, I thought that was a ClusterSetup method only.

Comment (michaelbeaumont) [reply test/e2e_env/universal/gateway/gateway.go:244]: If I'm ordering them I'll ust do the trafficpermission delete at the end so I can use one mesh.

Comment (jakubdyszkiewicz) [reply test/e2e_env/universal/gateway/gateway.go:244]: > Sure though a bit conflicted on making everything here order-dependent. My gut feeling says HTTP and HTTPS tests shouldn't depend on each other.

that's a good gut feeling! but don't they depend on the order of each other right now? Would this work if we run `when targeting a HTTPS gateway` first and then `when mTLS is disabled`?

Comment (michaelbeaumont) [reply test/e2e_env/universal/gateway/gateway.go:244]: In the first commit I don't think the order does matter. I use different meshes for mTLS enabled/disabled.

Comment (michaelbeaumont) [reply test/e2e_env/universal/gateway/gateway.go:227]: The changes I just pushed fully embraces `Ordered`.

File: test/e2e_env/universal/gateway/gateway.go (hunk 9 above)
Comment (michaelbeaumont) [reply test/e2e_env/universal/gateway/gateway.go:246]: I think the function is a bit more readable as is than converting the different calls into a `testCase struct` or interpreting arguments to the `DescribeTable` test function.

**Review by jakubdyszkiewicz — APPROVED** [review]

**Pull request merged**